Source/Services.Clients/ReverseCallClient.cs

```diff
@@ -0,0 +1,190 @@
+        /// <param name="setConnectArguments">A delegate to set the <typeparamref name="TConnectArguments" /> on a <typeparamref name="TClientMessage" />.</param>
+        /// <param name="getConnectResponse">A delegate to get the <typeparamref name="TConnectResponse" /> from a <typeparamref name="TServerMessage" />.</param>
+        /// <param name="getMessageRequest">A delegate to get the <typeparamref name="TRequest" /> from the <typeparamref name="TServerMessage" />.</param>
+        /// <param name="setMessageResponse">A delegate to set the <typeparamref name="TResponse" /> on a <typeparamref name="TClientMessage" />.</param>
+        /// <param name="setArgumentsContext">A delegate to set the <see cref="ReverseCallArgumentsContext" /> on a <typeparamref name="TConnectArguments" />.</param>
+        /// <param name="getRequestContext">A delegate to get the <see cref="ReverseCallRequestContext" /> from the <typeparamref name="TRequest" />.</param>
+        /// <param name="setResponseContext">A delegate to set the <see cref="ReverseCallResponseContext" /> on a <typeparamref name="TResponse" />.</param>
+        /// <param name="executionContextManager">The <see cref="IExecutionContextManager" />.</param>
+        /// <param name="logger">The <see cref="ILogger" />.</param>
+        public ReverseCallClient(
+            Func<AsyncDuplexStreamingCall<TClientMessage, TServerMessage>> establishConnection,
+            Action<TClientMessage, TConnectArguments> setConnectArguments,
+            Func<TServerMessage, TConnectResponse> getConnectResponse,
+            Func<TServerMessage, TRequest> getMessageRequest,
+            Action<TClientMessage, TResponse> setMessageResponse,
+            Action<TConnectArguments, ReverseCallArgumentsContext> setArgumentsContext,
+            Func<TRequest, ReverseCallRequestContext> getRequestContext,
+            Action<TResponse, ReverseCallResponseContext> setResponseContext,
+            IExecutionContextManager executionContextManager,
+            ILogger logger)
+        {
+            _establishConnection = establishConnection;
+            _setConnectArguments = setConnectArguments;
+            _getConnectResponse = getConnectResponse;
+            _getMessageRequest = getMessageRequest;
+            _setMessageResponse = setMessageResponse;
+            _setArgumentsContext = setArgumentsContext;
+            _getRequestContext = getRequestContext;
+            _setResponseContext = setResponseContext;
+            _executionContextManager = executionContextManager;
+            _logger = logger;
+        }
+
+        /// <inheritdoc/>
+        public TConnectResponse ConnectResponse { get; private set; }
+
+        /// <inheritdoc/>
+        public async Task<bool> Connect(TConnectArguments connectArguments, CancellationToken cancellationToken)
+        {
+            var streamingCall = _establishConnection();
```

Comment:
Same as below, this should only ever be called once.

Reply:
I think 0b1916ef2ca82aee2354e283f3f6e6d57459f234 resolves this

Reply:
Only thing I might want to change here is the name of the field. Maybe `_connecting` is a better name?

Reply:
Agree

Reply:
Fixed by https://github.com/dolittle-fundamentals/DotNET.Fundamentals/pull/310/commits/c6d51c68f18c7a09bbfb3c2796223eac69155b90